backend/src/openarchiefbeheer/zaken/utils.py

```diff
@@ -136,3 +145,141 @@ def retrieve_selectielijstklasse_choices(process_type_url: str) -> list:
         ]
 
     return results
+
+
+def execute_unless_result_exist(
+    store: ResultStore,
+    resource_type: str,
+    resource: str,
+    callable: Callable,
+    http_error_handler: Callable | None = None,
+) -> None:
+    if store.has_deleted_resource(resource_type, resource):
+        return
+
+    try:
+        response = callable(timeout=settings.REQUESTS_DEFAULT_TIMEOUT)
+        response.raise_for_status()
+    except HTTPError as exc:
+        if not http_error_handler:
+            raise exc
+
+        return http_error_handler(exc)
+
+    store.add_deleted_resource(resource_type, resource)
+    store.save()
+
+
+def delete_decisions(zaak: "Zaak", result_store: ResultStore) -> None:
+    brc_service = Service.objects.get(api_type=APITypes.brc)
+    brc_client = build_client(brc_service)
+
+    with brc_client:
+        response = brc_client.get("besluiten", params={"zaak": zaak.url})
+        response.raise_for_status()
+
+        data = response.json()
+        if data["count"] == 0:
+            return
+
+        for decision in data["results"]:
+            decision_uuid = furl(decision["url"]).path.segments[-1]
+            execute_unless_result_exist(
+                result_store,
+                "besluiten",
+                decision["url"],
+                partial(brc_client.delete, f"besluiten/{decision_uuid}"),
+            )
+
+
+def handle_document_delete_exc(exc: HTTPError) -> None:
+    response = exc.response
+    if not response.status_code == status.HTTP_204_NO_CONTENT and not (
+        response.status_code == status.HTTP_400_BAD_REQUEST
+        and response.json()["invalidParams"][0]["code"] == "pending-relations"
+    ):
+        response.raise_for_status()
+
+
+def delete_documents(
+    zaak: "Zaak", zrc_client: APIClient, result_store: ResultStore
+) -> None:
+    with zrc_client:
+        response = zrc_client.get("zaakinformatieobjecten", params={"zaak": zaak.url})
+        response.raise_for_status()
+        zios = response.json()
+        if not len(zios) and not result_store.has_resource_to_delete(
+            "enkelvoudiginformatieobjecten"
+        ):
+            return
+
+        for zio in zios:
+            result_store.add_resource_to_delete(
+                "enkelvoudiginformatieobjecten", zio["informatieobject"]
+            )
+            zio_uuid = furl(zio["url"]).path.segments[-1]
+            execute_unless_result_exist(
+                result_store,
+                "zaakinformatieobjecten",
+                zio["url"],
+                partial(zrc_client.delete, f"zaakinformatieobjecten/{zio_uuid}"),
+            )
+
+    drc_service = Service.objects.get(api_type=APITypes.drc)
+    drc_client = build_client(drc_service)
+
+    with drc_client:
+        for document_url in result_store.get_resources_to_delete(
+            "enkelvoudiginformatieobjecten"
+        ):
+            document_uuid = furl(document_url).path.segments[-1]
+            execute_unless_result_exist(
```

Comment:
I'm not sure if it's used, but theoretically it's possible that 1 document can be linked to several zaken.
So I think you need to check if the document is linked to zaken which are not planned to delete before destroying it

Reply:
This is indeed the case! If you try to delete a document that still has other relations, you get a 400 that says "pending-relations". So I decided to not do a GET and then a DELETE but just a DELETE and catch the 400 with that specific exception. :thinking: